# Xenium_benchmarking

This is a public repository for the following study available here:

Salas et al. **Deciphering tissue, inter- and intra-cellular transcriptional 
diversity with Xenium** (bioRxiv), 2022.

**Abstract**

Xenium is a new spatial transcriptomics product commercialized by 10X Genomics capable of generating maps of hundreds of transcripts in situ at a subcellular resolution. Herein, we explore  5 Xenium datasets of the mouse brain and  2 of human breast cancer by comparing scalability, resolution, data quality, capacities and limitations, with other spatially resolved transcriptomics (SRT) technologies. In addition, we benchmarked the performance of multiple open source computational tools when applied to Xenium datasets in tasks including cell segmentation, segmentation-free analysis, selection of spatially variable genes and domain identification, among others. To our knowledge, this study serves as the first independent analysis of the performance of Xenium.

***

## Datasets


***

## Folder structure


***

## Cloning and adding
In a clean conda environment with pip installed, run in the terminal:

```git clone https://github.com/Moldia/Xenium_benchmarking.git```

Navigate to the folder:

```cd Xenium_benchmarking```

And install using pip:

```pip install -e . ```
