{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### These jupyter can reproduced the results of figure2 in our papers.\n",
    "### you can use these functions we defined to evaluate the performance of each method with 10-X cross-validation.\n",
    "### If you want to use each method to analyse youe own data, please see the Tutorial.ipynb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Please note: We used raw count as input for predictions in this ipynb !!!\n",
    "\n",
    "### In our manuscripts, we tested four schemes of input expression matrices: \n",
    "### (1) raw expression matrix of spatial data and raw expression matrix of scRNA-seq data (R-R); \n",
    "### (2) normalized expression matrix of spatial data and raw expression matrix of scRNA-seq data (N-R); \n",
    "### (3) raw expression matrix of spatial data and normalized expression matrix of scRNA-seq data (R-N); \n",
    "### (4) normalized expression matrix of spatial data and normalized expression matrix of scRNA-seq data (N-N).\n",
    "\n",
    "### For N-N and N-R input, we calculate the accuracy by comapring the prediction result with normalized matrix of spatial data.\n",
    "### For R-R and R-N input, we calculate the accuracy by comapring the prediction result with raw expression matrix of spatial data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import sys\n",
    "import pickle\n",
    "import os\n",
    "import time as tm\n",
    "from functools import partial\n",
    "import scipy.stats as st\n",
    "from scipy.stats import wasserstein_distance\n",
    "import scipy.stats\n",
    "import copy\n",
    "from sklearn.model_selection import KFold\n",
    "import pandas as pd\n",
    "import multiprocessing\n",
    "import matplotlib as mpl \n",
    "import matplotlib.pyplot as plt\n",
    "import scanpy as sc\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "from scipy.spatial import distance_matrix\n",
    "from sklearn.metrics import matthews_corrcoef\n",
    "from scipy import stats\n",
    "import seaborn as sns\n",
    "\n",
    "from scipy.spatial.distance import cdist\n",
    "import h5py\n",
    "from scipy.stats import spearmanr\n",
    "\n",
    "import time\n",
    "import sys\n",
    "import tangram as tg\n",
    "from os.path import join\n",
    "from IPython.display import display\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def SpaGE_impute(K):\n",
    "    print ('We run SpaGE for this data\\n')\n",
    "    sys.path.append(\"Extenrnal/SpaGE-master/\")\n",
    "    from SpaGE.main import SpaGE\n",
    "    global RNA_data, Spatial_data, train_gene, predict_gene, iterations\n",
    "    \n",
    "    RNA_data = RNA_data.loc[(RNA_data.sum(axis=1) != 0)] # removes genes that are in the train list, causing error\n",
    "    RNA_data = RNA_data.loc[(RNA_data.var(axis=1) != 0)]\n",
    "        \n",
    "#     train_gene = []\n",
    "#     predict_gene = []\n",
    "\n",
    "#     for i in range(iterations):\n",
    "#         train_gene.append(list(RNA_data.index&Spatial_data.columns))\n",
    "#         predict_gene.append(list(set(RNA_data.index) - set(Spatial_data.columns))[:20])\n",
    "\n",
    "    # train = train_gene[K]\n",
    "    # predict = predict_gene[K]\n",
    "    \n",
    "    train = np.array(train_gene[K])\n",
    "    predict = np.array(predict_gene[K])\n",
    "        \n",
    "    print('train gene')\n",
    "\n",
    "    pv = len(train)/2\n",
    "    Spatial = Spatial_data[train]\n",
    "    Img_Genes = SpaGE(Spatial, RNA_data.T, n_pv = int(pv), genes_to_predict = predict)\n",
    "    \n",
    "    print('Img_Genes')\n",
    "    \n",
    "    result = Img_Genes[predict]\n",
    "    return result\n",
    "\n",
    "def gimVI_impute(K): # from tutorial\n",
    "    print ('We run gimVI for this data\\n')\n",
    "    import scvi\n",
    "    import scanpy as sc\n",
    "    from scvi.model import GIMVI\n",
    "    import torch\n",
    "    from torch.nn.functional import softmax, cosine_similarity, sigmoid\n",
    "    global RNA_data_adata, Spatial_data_adata, train_gene, predict_gene\n",
    "    \n",
    "    test_list = predict_gene[K]\n",
    "    train_list = train_gene[K]\n",
    "    \n",
    "    print('train list')\n",
    "\n",
    "    Genes  = train_list.copy()\n",
    "    Genes.extend(test_list)\n",
    "    rand_test_gene_idx = [Genes.index(x) for x in test_list]\n",
    "    n_genes = len(Genes)\n",
    "    rand_train_gene_idx = [Genes.index(x) for x in train_list]\n",
    "    rand_train_genes = np.array(Genes)[rand_train_gene_idx]\n",
    "    rand_test_genes = np.array(Genes)[rand_test_gene_idx]\n",
    "    spatial_data_partial = Spatial_data_adata[:, rand_train_genes]\n",
    "    sc.pp.filter_cells(spatial_data_partial, min_counts= 0)\n",
    "\n",
    "    seq_data = copy.deepcopy(RNA_data_adata)\n",
    "\n",
    "    seq_data = seq_data[:, Genes]\n",
    "    sc.pp.filter_cells(seq_data, min_counts = 0)\n",
    "    scvi.data.setup_anndata(spatial_data_partial)\n",
    "    scvi.data.setup_anndata(seq_data)\n",
    "\n",
    "    model = GIMVI(seq_data, spatial_data_partial)\n",
    "    model.train(200)\n",
    "\n",
    "    _, imputation = model.get_imputed_values(normalized=False)\n",
    "    imputed = imputation[:, rand_test_gene_idx]\n",
    "    result = pd.DataFrame(imputed, columns=rand_test_genes)\n",
    "\n",
    "    print (test_list ==  rand_test_genes)\n",
    "    print (train_list ==  rand_train_genes)\n",
    "    \n",
    "    return result\n",
    "                       \n",
    "def novoSpaRc_impute(K): # modified from tutorial\n",
    "    print ('We run novoSpaRc for this data\\n')\n",
    "    import novosparc as nc\n",
    "    global RNA_data, Spatial_data, locations, train_gene, predict_gene\n",
    "    \n",
    "    test_list = predict_gene[K]\n",
    "    train_list = train_gene[K]\n",
    "    \n",
    "    gene_names = np.array(RNA_data.index.values)\n",
    "    dge = RNA_data.values\n",
    "    dge = dge.T\n",
    "    num_cells = dge.shape[0]\n",
    "    print ('number of cells and genes in the matrix:', dge.shape)\n",
    "\n",
    "    hvg = np.argsort(np.divide(np.var(dge,axis=0),np.mean(dge,axis=0)+0.0001))\n",
    "    dge_hvg = dge[:,hvg[-2000:]]\n",
    "        \n",
    "    num_locations = locations.shape[0]\n",
    "    \n",
    "    print('num locations: ', num_locations)\n",
    "        \n",
    "    p_location, p_expression = nc.rc.create_space_distributions(num_locations, num_cells)\n",
    "    cost_expression, cost_locations = nc.rc.setup_for_OT_reconstruction(dge_hvg,locations,num_neighbors_source = 5,num_neighbors_target = 5)\n",
    "\n",
    "    insitu_matrix = np.array(Spatial_data[train_list])\n",
    "    insitu_genes = np.array(Spatial_data[train_list].columns)\n",
    "    test_genes = np.array(test_list)\n",
    "        \n",
    "    markers_in_sc = np.array([], dtype='int')\n",
    "    for marker in insitu_genes:\n",
    "        marker_index = np.where(gene_names == marker)[0]\n",
    "        if len(marker_index) > 0:\n",
    "            markers_in_sc = np.append(markers_in_sc, marker_index[0])\n",
    "\n",
    "    cost_marker_genes = cdist(dge[:, markers_in_sc]/np.amax(dge[:, markers_in_sc]),insitu_matrix/np.amax(insitu_matrix))\n",
    "    alpha_linear = 0.5\n",
    "        \n",
    "    gw = nc.rc._GWadjusted.gromov_wasserstein_adjusted_norm(cost_marker_genes, cost_expression, cost_locations,alpha_linear, p_expression, p_location,'square_loss', epsilon=5e-3, verbose=True)\n",
    "\n",
    "    sdge = np.dot(dge.T, gw)\n",
    "    imputed = pd.DataFrame(sdge,index=RNA_data.index)\n",
    "    result = imputed.loc[test_genes]\n",
    "    result = result.T\n",
    "    \n",
    "    return result\n",
    "   \n",
    "                       \n",
    "def SpaOTsc_impute(K):\n",
    "    print ('We run SpaOTsc for this data\\n')\n",
    "    from spaotsc import SpaOTsc\n",
    "    global RNA_data, Spatial_data, locations, train_gene, predict_gene\n",
    "    \n",
    "    test_list = predict_gene[K]\n",
    "    train_list = train_gene[K]\n",
    "\n",
    "    print('train list')\n",
    "                          \n",
    "    df_sc = RNA_data.T\n",
    "    df_IS = Spatial_data\n",
    "    pts = locations\n",
    "    is_dmat = distance_matrix(pts, pts)\n",
    "    df_is = df_IS.loc[:,train_list]\n",
    "    \n",
    "    print('data')\n",
    "    \n",
    "    gene_is = df_is.columns.tolist()\n",
    "    gene_sc = df_sc.columns.tolist()\n",
    "    gene_overloap = list(set(gene_is).intersection(gene_sc))\n",
    "    a = df_is[gene_overloap]\n",
    "    b = df_sc[gene_overloap]\n",
    "    \n",
    "    print('gene')\n",
    "    \n",
    "    rho, pval = stats.spearmanr(a, b,axis=1)\n",
    "    rho[np.isnan(rho)]=0\n",
    "    mcc=rho[-(len(df_sc)):,0:len(df_is)]\n",
    "    C = np.exp(1 - mcc)\n",
    "        \n",
    "    issc = SpaOTsc.spatial_sc(sc_data = df_sc, is_data = df_is, is_dmat = is_dmat)\n",
    "    issc.transport_plan(C**2, alpha = 0, rho = 1.0, epsilon = 0.1, cor_matrix = mcc, scaling = False)\n",
    "        \n",
    "    print('spatial')\n",
    "        \n",
    "    gamma = issc.gamma_mapping\n",
    "    for j in range(gamma.shape[1]):\n",
    "        gamma[:,j] = gamma[:,j]/np.sum(gamma[:,j])\n",
    "    X_pred = np.matmul(gamma.T, np.array(issc.sc_data.values))\n",
    "        \n",
    "    result = pd.DataFrame(data = X_pred, columns = issc.sc_data.columns.values)\n",
    "    test_genes = test_list\n",
    "    result = result.loc[:, test_genes]\n",
    "    return result\n",
    "\n",
    "def stPlus_impute(K):\n",
    "    global RNA_data, Spatial_data, outdir, train_gene, predict_gene\n",
    "    \n",
    "    test_list = predict_gene[K]\n",
    "    train_list = train_gene[K]\n",
    "    \n",
    "    save_path_prefix = join(outdir, 'process_file/stPlus-demo')\n",
    "    if not os.path.exists(join(outdir, \"process_file\")):\n",
    "        os.mkdir(join(outdir, \"process_file\"))\n",
    "    stPlus_res = stPlus(Spatial_data[train_list], RNA_data.T, test_list, save_path_prefix)\n",
    "    \n",
    "    return stPlus_res\n",
    "\n",
    "\n",
    "def Tangram_impute(K, annotate = None, modes = 'clusters', density = 'rna_count_based'): # modified from tutorial\n",
    "    import torch\n",
    "    from torch.nn.functional import softmax, cosine_similarity, sigmoid\n",
    "    import tangram as tg\n",
    "    print ('We run Tangram for this \\n' + DataDir)\n",
    "    global RNA_data_adata, Spatial_data_adata, locations, train_gene, predict_gene\n",
    "    \n",
    "\n",
    "    test_list = predict_gene[K]\n",
    "    test_list = [x.lower() for x in test_list]\n",
    "    train_list = train_gene[K]\n",
    "    \n",
    "    if annotate == None:\n",
    "        RNA_data_adata_label = RNA_data_adata\n",
    "        sc.pp.normalize_total(RNA_data_adata_label)\n",
    "        sc.pp.log1p(RNA_data_adata_label)\n",
    "        sc.pp.highly_variable_genes(RNA_data_adata_label)\n",
    "        RNA_data_adata_label = RNA_data_adata_label[:, RNA_data_adata_label.var.highly_variable]\n",
    "        sc.pp.scale(RNA_data_adata_label, max_value=10)\n",
    "        sc.tl.pca(RNA_data_adata_label)\n",
    "        sc.pp.neighbors(RNA_data_adata_label)\n",
    "        sc.tl.leiden(RNA_data_adata_label, resolution = 0.5)\n",
    "        RNA_data_adata.obs['leiden']  = RNA_data_adata_label.obs.leiden\n",
    "        tg.pp_adatas(RNA_data_adata, Spatial_data_adata, genes=train_list)\n",
    "    else:\n",
    "        global CellTypeAnnotate \n",
    "        RNA_data_adata.obs['leiden']  = CellTypeAnnotate\n",
    "        tg.pp_adatas(RNA_data_adata, Spatial_data_adata, genes=train_list)\n",
    "        \n",
    "    device = torch.device('cuda:0')\n",
    "    \n",
    "    if modes == 'clusters':\n",
    "        ad_map = tg.map_cells_to_space(RNA_data_adata, Spatial_data_adata, device = device, mode = modes, cluster_label = 'leiden', density_prior = density)\n",
    "        ad_ge = tg.project_genes(ad_map, RNA_data_adata, cluster_label = 'leiden')\n",
    "    else:\n",
    "        ad_map = tg.map_cells_to_space(RNA_data_adata, Spatial_data_adata, device = device)\n",
    "        ad_ge = tg.project_genes(ad_map, RNA_data_adata)\n",
    "        \n",
    "    test_list = list(set(ad_ge.var_names) & set(test_list))\n",
    "    test_list = np.array(test_list)\n",
    "    pre_gene = pd.DataFrame(ad_ge[:,test_list].X, index=ad_ge[:,test_list].obs_names, columns=ad_ge[:,test_list].var_names)\n",
    "    \n",
    "    return pre_gene\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Input"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# First, you can download Example datasets from our websites "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "@author: wen zhang\n",
    "This function integrates two single-cell datasets, spatial and scRNA-seq, \n",
    "and predictes the expression of the spatially unmeasured genes from the scRNA-seq data.\n",
    "\n",
    "Parameters\n",
    "-------\n",
    "RNA_file : str\n",
    "    scRNA-seq data count file with Tab-delimited (cells X genes).\n",
    "Spatial_file : str\n",
    "    spatial count data file with Tab-delimited, please note that the file has no index.\n",
    "location_file : str\n",
    "    spatial spot coordinate file name with Tab-delimited, please note that the file has no index.\n",
    "device : str\n",
    "    Option,  ['CPU','GPU'], defaults to 'CPU'\n",
    "train_gene : list\n",
    "    genes for integrations, you can support more than one train list.\n",
    "predict_gene : list\n",
    "    genes for prediction, you can support more than one test list.\n",
    "outdir : str\n",
    "    result file stored direction    \n",
    "\"\"\"\n",
    "\n",
    "DataDir = 'xenium data/'\n",
    "\n",
    "RNA_file = DataDir + 'Yao_150kcells_subsample_with_annotations.h5ad'\n",
    "Spatial_file = DataDir + 'spatial_data.csv'\n",
    "location_file = DataDir + 'locations.csv'\n",
    "\n",
    "\n",
    "device = 'GPU'\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['spage', 'spaotsc', 'novosparc', 'gimvi', 'tangram']\n"
     ]
    }
   ],
   "source": [
    "# test_cells = sys.argv[1]\n",
    "# iterations = sys.argv[2]\n",
    "# fraction = sys.argv[3]\n",
    "# methods = list(sys.argv[4:])\n",
    "\n",
    "methods = ['spage', 'spaotsc', 'novosparc', 'gimvi', 'tangram']\n",
    "# methods = ['gimvi', 'tangram']\n",
    "\n",
    "\n",
    "print(methods)\n",
    "\n",
    "test_cells = 10\n",
    "iterations = 10\n",
    "fraction = 10\n",
    "iterations = int(iterations)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "outdir = 'xenium data/Figure2'\n",
    "\n",
    "if not os.path.exists(outdir):\n",
    "    os.mkdir(outdir)\n",
    "\n",
    "outdir = outdir + '/downsampling_' + str(test_cells) \n",
    "\n",
    "if not os.path.exists(outdir):\n",
    "    os.mkdir(outdir)\n",
    "    \n",
    "outdir = outdir + '/iterations_' + str(iterations)\n",
    "\n",
    "if not os.path.exists(outdir):\n",
    "    os.mkdir(outdir)\n",
    "    \n",
    "outdir = outdir + '/split_' + str(fraction)\n",
    "\n",
    "if not os.path.exists(outdir):\n",
    "    os.mkdir(outdir)\n",
    "    \n",
    "outdir = outdir + '/impute'\n",
    "\n",
    "if not os.path.exists(outdir):\n",
    "    os.mkdir(outdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.1\n"
     ]
    }
   ],
   "source": [
    "# test_cells = 0.01\n",
    "\n",
    "test_cells = 1/ int(test_cells)\n",
    "print(test_cells)\n",
    "\n",
    "rna_cells = int(149955 * test_cells)\n",
    "spatial_spots = int(82941 *test_cells)\n",
    "\n",
    "# rna_cells=1000\n",
    "# spatial_spots=1000"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### RNA data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 149955 × 31053\n",
       "    obs: 'sample_name', 'Unnamed: 0', 'cl', 'exp_component_name', 'donor_label', 'sex_label', 'region_label', 'gene.counts', 'library_label', 'platform_label', 'joint_region_color', 'joint_region_id', 'joint_region_label', 'region_id', 'region_color', 'sex_color', 'sex_id', 'platform_color', 'platform_id', 'tasic18_cluster_id', 'tasic18_cluster_label', 'tasic18_cluster_color', 'tasic18_subclass_id', 'tasic18_subclass_label', 'tasic18_subclass_color', 'tasic18_class_id', 'tasic18_class_label', 'tasic18_class_color', 'Miniatlas_cluster_id', 'Miniatlas_cluster_label', 'Miniatlas_cluster_color', 'genes_label', 'genes_norm', 'genes_color', 'cluster_id', 'cluster_label', 'ref.cluster_label', 'old.cluster_label', 'label', 'cluster_color', 'supertype_id', 'supertype_label', 'supertype_color', 'subclass_id', 'subclass_label', 'subclass_color', 'neighborhood_id', 'neighborhood_label', 'neighborhood_color', 'class_id', 'class_label', 'class_color', 'ss_cluster_id', 'ss_cluster_color', 'ss_cluster_label', 'tenx_cluster_id', 'tenx_cluster_color', 'tenx_cluster_label', 'cell_type_accession_id', 'cell_type_accession_color', 'cell_type_accession_label', 'sss'\n",
       "    var: 'gene'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RNA_data_adata = sc.read_h5ad(RNA_file)\n",
    "RNA_data_adata\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Xkr4</th>\n",
       "      <th>Gm1992</th>\n",
       "      <th>Gm37381</th>\n",
       "      <th>Rp1</th>\n",
       "      <th>Sox17</th>\n",
       "      <th>Gm37323</th>\n",
       "      <th>Mrpl15</th>\n",
       "      <th>Lypla1</th>\n",
       "      <th>Gm37988</th>\n",
       "      <th>Tcea1</th>\n",
       "      <th>...</th>\n",
       "      <th>AC125149.1</th>\n",
       "      <th>AC125149.2</th>\n",
       "      <th>AC125149.4</th>\n",
       "      <th>AC234645.1</th>\n",
       "      <th>AC168977.2</th>\n",
       "      <th>AC168977.1</th>\n",
       "      <th>AC149090.1</th>\n",
       "      <th>CAAA01118383.1</th>\n",
       "      <th>Vmn2r122</th>\n",
       "      <th>CAAA01147332.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10X_cells.AAACCTGAGCCTTGAT-L8TX_180221_01_F09</th>\n",
       "      <td>11.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10X_cells.AAACCTGCAAGCTGTT-L8TX_180221_01_F09</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10X_cells.AAACCTGCACGTCTCT-L8TX_180221_01_F09</th>\n",
       "      <td>30.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10X_cells.AAACCTGGTTGTCGCG-L8TX_180221_01_F09</th>\n",
       "      <td>14.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10X_cells.AAACCTGGTTTCGCTC-L8TX_180221_01_F09</th>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10X_cells.TTTATGCCAAGGTTCT-L8TX_200611_02_A05</th>\n",
       "      <td>14.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10X_cells.TTTGCGCAGCTGAAAT-L8TX_200611_02_A05</th>\n",
       "      <td>18.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10X_cells.TTTGGTTAGGGTTTCT-L8TX_200611_02_A05</th>\n",
       "      <td>12.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10X_cells.TTTGGTTCACCGATAT-L8TX_200611_02_A05</th>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10X_cells.TTTGTCATCGTCTGAA-L8TX_200611_02_A05</th>\n",
       "      <td>16.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>149955 rows × 31053 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               Xkr4  Gm1992  Gm37381  Rp1  \\\n",
       "10X_cells.AAACCTGAGCCTTGAT-L8TX_180221_01_F09  11.0     0.0      0.0  0.0   \n",
       "10X_cells.AAACCTGCAAGCTGTT-L8TX_180221_01_F09   0.0     0.0      0.0  0.0   \n",
       "10X_cells.AAACCTGCACGTCTCT-L8TX_180221_01_F09  30.0     0.0      0.0  0.0   \n",
       "10X_cells.AAACCTGGTTGTCGCG-L8TX_180221_01_F09  14.0     0.0      0.0  0.0   \n",
       "10X_cells.AAACCTGGTTTCGCTC-L8TX_180221_01_F09  15.0     0.0      0.0  0.0   \n",
       "...                                             ...     ...      ...  ...   \n",
       "10X_cells.TTTATGCCAAGGTTCT-L8TX_200611_02_A05  14.0     1.0      0.0  0.0   \n",
       "10X_cells.TTTGCGCAGCTGAAAT-L8TX_200611_02_A05  18.0     2.0      0.0  0.0   \n",
       "10X_cells.TTTGGTTAGGGTTTCT-L8TX_200611_02_A05  12.0     1.0      0.0  0.0   \n",
       "10X_cells.TTTGGTTCACCGATAT-L8TX_200611_02_A05   7.0     0.0      0.0  0.0   \n",
       "10X_cells.TTTGTCATCGTCTGAA-L8TX_200611_02_A05  16.0     1.0      0.0  0.0   \n",
       "\n",
       "                                               Sox17  Gm37323  Mrpl15  Lypla1  \\\n",
       "10X_cells.AAACCTGAGCCTTGAT-L8TX_180221_01_F09    0.0      0.0     0.0     0.0   \n",
       "10X_cells.AAACCTGCAAGCTGTT-L8TX_180221_01_F09    0.0      0.0     0.0     0.0   \n",
       "10X_cells.AAACCTGCACGTCTCT-L8TX_180221_01_F09    0.0      0.0     0.0     0.0   \n",
       "10X_cells.AAACCTGGTTGTCGCG-L8TX_180221_01_F09    0.0      0.0     1.0     0.0   \n",
       "10X_cells.AAACCTGGTTTCGCTC-L8TX_180221_01_F09    0.0      0.0     0.0     0.0   \n",
       "...                                              ...      ...     ...     ...   \n",
       "10X_cells.TTTATGCCAAGGTTCT-L8TX_200611_02_A05    0.0      0.0     0.0     0.0   \n",
       "10X_cells.TTTGCGCAGCTGAAAT-L8TX_200611_02_A05    0.0      0.0     0.0     0.0   \n",
       "10X_cells.TTTGGTTAGGGTTTCT-L8TX_200611_02_A05    0.0      0.0     0.0     0.0   \n",
       "10X_cells.TTTGGTTCACCGATAT-L8TX_200611_02_A05    0.0      0.0     0.0     0.0   \n",
       "10X_cells.TTTGTCATCGTCTGAA-L8TX_200611_02_A05    0.0      0.0     1.0     0.0   \n",
       "\n",
       "                                               Gm37988  Tcea1  ...  \\\n",
       "10X_cells.AAACCTGAGCCTTGAT-L8TX_180221_01_F09      0.0    1.0  ...   \n",
       "10X_cells.AAACCTGCAAGCTGTT-L8TX_180221_01_F09      0.0    0.0  ...   \n",
       "10X_cells.AAACCTGCACGTCTCT-L8TX_180221_01_F09      0.0    0.0  ...   \n",
       "10X_cells.AAACCTGGTTGTCGCG-L8TX_180221_01_F09      0.0    2.0  ...   \n",
       "10X_cells.AAACCTGGTTTCGCTC-L8TX_180221_01_F09      0.0    0.0  ...   \n",
       "...                                                ...    ...  ...   \n",
       "10X_cells.TTTATGCCAAGGTTCT-L8TX_200611_02_A05      1.0    0.0  ...   \n",
       "10X_cells.TTTGCGCAGCTGAAAT-L8TX_200611_02_A05      0.0    1.0  ...   \n",
       "10X_cells.TTTGGTTAGGGTTTCT-L8TX_200611_02_A05      0.0    0.0  ...   \n",
       "10X_cells.TTTGGTTCACCGATAT-L8TX_200611_02_A05      0.0    0.0  ...   \n",
       "10X_cells.TTTGTCATCGTCTGAA-L8TX_200611_02_A05      0.0    1.0  ...   \n",
       "\n",
       "                                               AC125149.1  AC125149.2  \\\n",
       "10X_cells.AAACCTGAGCCTTGAT-L8TX_180221_01_F09         0.0         0.0   \n",
       "10X_cells.AAACCTGCAAGCTGTT-L8TX_180221_01_F09         0.0         0.0   \n",
       "10X_cells.AAACCTGCACGTCTCT-L8TX_180221_01_F09         0.0         0.0   \n",
       "10X_cells.AAACCTGGTTGTCGCG-L8TX_180221_01_F09         0.0         0.0   \n",
       "10X_cells.AAACCTGGTTTCGCTC-L8TX_180221_01_F09         0.0         0.0   \n",
       "...                                                   ...         ...   \n",
       "10X_cells.TTTATGCCAAGGTTCT-L8TX_200611_02_A05         0.0         0.0   \n",
       "10X_cells.TTTGCGCAGCTGAAAT-L8TX_200611_02_A05         0.0         0.0   \n",
       "10X_cells.TTTGGTTAGGGTTTCT-L8TX_200611_02_A05         0.0         0.0   \n",
       "10X_cells.TTTGGTTCACCGATAT-L8TX_200611_02_A05         0.0         0.0   \n",
       "10X_cells.TTTGTCATCGTCTGAA-L8TX_200611_02_A05         0.0         0.0   \n",
       "\n",
       "                                               AC125149.4  AC234645.1  \\\n",
       "10X_cells.AAACCTGAGCCTTGAT-L8TX_180221_01_F09         0.0         0.0   \n",
       "10X_cells.AAACCTGCAAGCTGTT-L8TX_180221_01_F09         0.0         0.0   \n",
       "10X_cells.AAACCTGCACGTCTCT-L8TX_180221_01_F09         0.0         0.0   \n",
       "10X_cells.AAACCTGGTTGTCGCG-L8TX_180221_01_F09         0.0         0.0   \n",
       "10X_cells.AAACCTGGTTTCGCTC-L8TX_180221_01_F09         0.0         0.0   \n",
       "...                                                   ...         ...   \n",
       "10X_cells.TTTATGCCAAGGTTCT-L8TX_200611_02_A05         0.0         0.0   \n",
       "10X_cells.TTTGCGCAGCTGAAAT-L8TX_200611_02_A05         0.0         0.0   \n",
       "10X_cells.TTTGGTTAGGGTTTCT-L8TX_200611_02_A05         0.0         0.0   \n",
       "10X_cells.TTTGGTTCACCGATAT-L8TX_200611_02_A05         0.0         0.0   \n",
       "10X_cells.TTTGTCATCGTCTGAA-L8TX_200611_02_A05         0.0         0.0   \n",
       "\n",
       "                                               AC168977.2  AC168977.1  \\\n",
       "10X_cells.AAACCTGAGCCTTGAT-L8TX_180221_01_F09         0.0         0.0   \n",
       "10X_cells.AAACCTGCAAGCTGTT-L8TX_180221_01_F09         0.0         0.0   \n",
       "10X_cells.AAACCTGCACGTCTCT-L8TX_180221_01_F09         0.0         0.0   \n",
       "10X_cells.AAACCTGGTTGTCGCG-L8TX_180221_01_F09         0.0         0.0   \n",
       "10X_cells.AAACCTGGTTTCGCTC-L8TX_180221_01_F09         0.0         0.0   \n",
       "...                                                   ...         ...   \n",
       "10X_cells.TTTATGCCAAGGTTCT-L8TX_200611_02_A05         0.0         0.0   \n",
       "10X_cells.TTTGCGCAGCTGAAAT-L8TX_200611_02_A05         0.0         0.0   \n",
       "10X_cells.TTTGGTTAGGGTTTCT-L8TX_200611_02_A05         0.0         0.0   \n",
       "10X_cells.TTTGGTTCACCGATAT-L8TX_200611_02_A05         0.0         0.0   \n",
       "10X_cells.TTTGTCATCGTCTGAA-L8TX_200611_02_A05         0.0         0.0   \n",
       "\n",
       "                                               AC149090.1  CAAA01118383.1  \\\n",
       "10X_cells.AAACCTGAGCCTTGAT-L8TX_180221_01_F09         1.0             0.0   \n",
       "10X_cells.AAACCTGCAAGCTGTT-L8TX_180221_01_F09         1.0             0.0   \n",
       "10X_cells.AAACCTGCACGTCTCT-L8TX_180221_01_F09         0.0             0.0   \n",
       "10X_cells.AAACCTGGTTGTCGCG-L8TX_180221_01_F09         0.0             0.0   \n",
       "10X_cells.AAACCTGGTTTCGCTC-L8TX_180221_01_F09         2.0             0.0   \n",
       "...                                                   ...             ...   \n",
       "10X_cells.TTTATGCCAAGGTTCT-L8TX_200611_02_A05         5.0             0.0   \n",
       "10X_cells.TTTGCGCAGCTGAAAT-L8TX_200611_02_A05         1.0             0.0   \n",
       "10X_cells.TTTGGTTAGGGTTTCT-L8TX_200611_02_A05         7.0             0.0   \n",
       "10X_cells.TTTGGTTCACCGATAT-L8TX_200611_02_A05         2.0             0.0   \n",
       "10X_cells.TTTGTCATCGTCTGAA-L8TX_200611_02_A05         5.0             0.0   \n",
       "\n",
       "                                               Vmn2r122  CAAA01147332.1  \n",
       "10X_cells.AAACCTGAGCCTTGAT-L8TX_180221_01_F09       0.0             0.0  \n",
       "10X_cells.AAACCTGCAAGCTGTT-L8TX_180221_01_F09       0.0             0.0  \n",
       "10X_cells.AAACCTGCACGTCTCT-L8TX_180221_01_F09       0.0             0.0  \n",
       "10X_cells.AAACCTGGTTGTCGCG-L8TX_180221_01_F09       0.0             0.0  \n",
       "10X_cells.AAACCTGGTTTCGCTC-L8TX_180221_01_F09       0.0             0.0  \n",
       "...                                                 ...             ...  \n",
       "10X_cells.TTTATGCCAAGGTTCT-L8TX_200611_02_A05       0.0             0.0  \n",
       "10X_cells.TTTGCGCAGCTGAAAT-L8TX_200611_02_A05       0.0             0.0  \n",
       "10X_cells.TTTGGTTAGGGTTTCT-L8TX_200611_02_A05       0.0             0.0  \n",
       "10X_cells.TTTGGTTCACCGATAT-L8TX_200611_02_A05       0.0             0.0  \n",
       "10X_cells.TTTGTCATCGTCTGAA-L8TX_200611_02_A05       0.0             0.0  \n",
       "\n",
       "[149955 rows x 31053 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RNA_data = pd.DataFrame(RNA_data_adata.X, columns = list(RNA_data_adata.var.index), index = list(RNA_data_adata.obs['sample_name']))\n",
    "RNA_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>10X_cells.ATAAGAGCATACGCCG-L8TX_180221_01_B10</th>\n",
       "      <th>10X_cells.AGATTGCTCACATACG-L8TX_180221_01_A10</th>\n",
       "      <th>10X_cells.TGAGGGATCGGCTACG-L8TX_180221_01_C10</th>\n",
       "      <th>10X_cells.AGGCCACCAGCGAACA-L8TX_180221_01_D09</th>\n",
       "      <th>10X_cells.GTTCTCGTCTGTTTGT-L8TX_180221_01_F12</th>\n",
       "      <th>10X_cells.TGAGAGGGTGTTTGTG-L8TX_171026_01_C05</th>\n",
       "      <th>10X_cells.GGTATTGGTTTAAGCC-L8TX_171026_01_H03</th>\n",
       "      <th>10X_cells.TTGCGTCTCGGTCCGA-L8TX_180221_01_E12</th>\n",
       "      <th>10X_cells.CGAGAAGGTTGTGGAG-L8TX_181011_01_A03</th>\n",
       "      <th>10X_cells.CACAGGCAGCTGATAA-L8TX_180926_01_F01</th>\n",
       "      <th>...</th>\n",
       "      <th>10X_cells.CCATGTCAGTTACGGG-L8TX_171026_01_E05</th>\n",
       "      <th>10X_cells.CTTTGCGTCATGGTCA-L8TX_180115_01_F11</th>\n",
       "      <th>10X_cells.GGAAAGCGTGCAACGA-L8TX_171026_01_H04</th>\n",
       "      <th>10X_cells.AAGACCTAGTGGCACA-L8TX_190312_01_C02</th>\n",
       "      <th>10X_cells.CCACGGACATCGTCGG-L8TX_180406_01_C06</th>\n",
       "      <th>10X_cells.GAAGCAGCAGCGATCC-L8TX_180221_01_C09</th>\n",
       "      <th>10X_cells.TCAGGTACAGCGTAAG-L8TX_180926_01_C12</th>\n",
       "      <th>10X_cells.CGGACGTCACATGTGT-L8TX_180115_01_G11</th>\n",
       "      <th>10X_cells.TACTTGTCACATTAGC-L8TX_180221_01_E10</th>\n",
       "      <th>10X_cells.CCTTACGGTTCGTGAT-L8TX_180115_01_H10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Xkr4</th>\n",
       "      <td>7.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>10.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gm1992</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gm37381</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rp1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sox17</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC168977.1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC149090.1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAAA01118383.1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Vmn2r122</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAAA01147332.1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>31053 rows × 14995 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                10X_cells.ATAAGAGCATACGCCG-L8TX_180221_01_B10  \\\n",
       "Xkr4                                                      7.0   \n",
       "Gm1992                                                    0.0   \n",
       "Gm37381                                                   0.0   \n",
       "Rp1                                                       0.0   \n",
       "Sox17                                                     0.0   \n",
       "...                                                       ...   \n",
       "AC168977.1                                                0.0   \n",
       "AC149090.1                                                0.0   \n",
       "CAAA01118383.1                                            0.0   \n",
       "Vmn2r122                                                  0.0   \n",
       "CAAA01147332.1                                            0.0   \n",
       "\n",
       "                10X_cells.AGATTGCTCACATACG-L8TX_180221_01_A10  \\\n",
       "Xkr4                                                      3.0   \n",
       "Gm1992                                                    0.0   \n",
       "Gm37381                                                   0.0   \n",
       "Rp1                                                       0.0   \n",
       "Sox17                                                     0.0   \n",
       "...                                                       ...   \n",
       "AC168977.1                                                0.0   \n",
       "AC149090.1                                                0.0   \n",
       "CAAA01118383.1                                            0.0   \n",
       "Vmn2r122                                                  0.0   \n",
       "CAAA01147332.1                                            0.0   \n",
       "\n",
       "                10X_cells.TGAGGGATCGGCTACG-L8TX_180221_01_C10  \\\n",
       "Xkr4                                                      4.0   \n",
       "Gm1992                                                    0.0   \n",
       "Gm37381                                                   0.0   \n",
       "Rp1                                                       0.0   \n",
       "Sox17                                                     0.0   \n",
       "...                                                       ...   \n",
       "AC168977.1                                                0.0   \n",
       "AC149090.1                                                2.0   \n",
       "CAAA01118383.1                                            0.0   \n",
       "Vmn2r122                                                  0.0   \n",
       "CAAA01147332.1                                            0.0   \n",
       "\n",
       "                10X_cells.AGGCCACCAGCGAACA-L8TX_180221_01_D09  \\\n",
       "Xkr4                                                     15.0   \n",
       "Gm1992                                                    0.0   \n",
       "Gm37381                                                   0.0   \n",
       "Rp1                                                       0.0   \n",
       "Sox17                                                     0.0   \n",
       "...                                                       ...   \n",
       "AC168977.1                                                0.0   \n",
       "AC149090.1                                                1.0   \n",
       "CAAA01118383.1                                            0.0   \n",
       "Vmn2r122                                                  0.0   \n",
       "CAAA01147332.1                                            0.0   \n",
       "\n",
       "                10X_cells.GTTCTCGTCTGTTTGT-L8TX_180221_01_F12  \\\n",
       "Xkr4                                                      3.0   \n",
       "Gm1992                                                    0.0   \n",
       "Gm37381                                                   0.0   \n",
       "Rp1                                                       0.0   \n",
       "Sox17                                                     0.0   \n",
       "...                                                       ...   \n",
       "AC168977.1                                                0.0   \n",
       "AC149090.1                                                0.0   \n",
       "CAAA01118383.1                                            0.0   \n",
       "Vmn2r122                                                  0.0   \n",
       "CAAA01147332.1                                            0.0   \n",
       "\n",
       "                10X_cells.TGAGAGGGTGTTTGTG-L8TX_171026_01_C05  \\\n",
       "Xkr4                                                      8.0   \n",
       "Gm1992                                                    1.0   \n",
       "Gm37381                                                   0.0   \n",
       "Rp1                                                       0.0   \n",
       "Sox17                                                     0.0   \n",
       "...                                                       ...   \n",
       "AC168977.1                                                0.0   \n",
       "AC149090.1                                                0.0   \n",
       "CAAA01118383.1                                            0.0   \n",
       "Vmn2r122                                                  0.0   \n",
       "CAAA01147332.1                                            0.0   \n",
       "\n",
       "                10X_cells.GGTATTGGTTTAAGCC-L8TX_171026_01_H03  \\\n",
       "Xkr4                                                      8.0   \n",
       "Gm1992                                                    0.0   \n",
       "Gm37381                                                   0.0   \n",
       "Rp1                                                       0.0   \n",
       "Sox17                                                     0.0   \n",
       "...                                                       ...   \n",
       "AC168977.1                                                0.0   \n",
       "AC149090.1                                                1.0   \n",
       "CAAA01118383.1                                            0.0   \n",
       "Vmn2r122                                                  0.0   \n",
       "CAAA01147332.1                                            0.0   \n",
       "\n",
       "                10X_cells.TTGCGTCTCGGTCCGA-L8TX_180221_01_E12  \\\n",
       "Xkr4                                                      8.0   \n",
       "Gm1992                                                    0.0   \n",
       "Gm37381                                                   0.0   \n",
       "Rp1                                                       0.0   \n",
       "Sox17                                                     0.0   \n",
       "...                                                       ...   \n",
       "AC168977.1                                                0.0   \n",
       "AC149090.1                                                0.0   \n",
       "CAAA01118383.1                                            0.0   \n",
       "Vmn2r122                                                  0.0   \n",
       "CAAA01147332.1                                            0.0   \n",
       "\n",
       "                10X_cells.CGAGAAGGTTGTGGAG-L8TX_181011_01_A03  \\\n",
       "Xkr4                                                      5.0   \n",
       "Gm1992                                                    0.0   \n",
       "Gm37381                                                   0.0   \n",
       "Rp1                                                       0.0   \n",
       "Sox17                                                     0.0   \n",
       "...                                                       ...   \n",
       "AC168977.1                                                0.0   \n",
       "AC149090.1                                                7.0   \n",
       "CAAA01118383.1                                            0.0   \n",
       "Vmn2r122                                                  0.0   \n",
       "CAAA01147332.1                                            0.0   \n",
       "\n",
       "                10X_cells.CACAGGCAGCTGATAA-L8TX_180926_01_F01  ...  \\\n",
       "Xkr4                                                      0.0  ...   \n",
       "Gm1992                                                    0.0  ...   \n",
       "Gm37381                                                   0.0  ...   \n",
       "Rp1                                                       0.0  ...   \n",
       "Sox17                                                     0.0  ...   \n",
       "...                                                       ...  ...   \n",
       "AC168977.1                                                0.0  ...   \n",
       "AC149090.1                                                0.0  ...   \n",
       "CAAA01118383.1                                            0.0  ...   \n",
       "Vmn2r122                                                  0.0  ...   \n",
       "CAAA01147332.1                                            0.0  ...   \n",
       "\n",
       "                10X_cells.CCATGTCAGTTACGGG-L8TX_171026_01_E05  \\\n",
       "Xkr4                                                     10.0   \n",
       "Gm1992                                                    0.0   \n",
       "Gm37381                                                   0.0   \n",
       "Rp1                                                       0.0   \n",
       "Sox17                                                     0.0   \n",
       "...                                                       ...   \n",
       "AC168977.1                                                0.0   \n",
       "AC149090.1                                                3.0   \n",
       "CAAA01118383.1                                            0.0   \n",
       "Vmn2r122                                                  0.0   \n",
       "CAAA01147332.1                                            0.0   \n",
       "\n",
       "                10X_cells.CTTTGCGTCATGGTCA-L8TX_180115_01_F11  \\\n",
       "Xkr4                                                      3.0   \n",
       "Gm1992                                                    0.0   \n",
       "Gm37381                                                   0.0   \n",
       "Rp1                                                       0.0   \n",
       "Sox17                                                     0.0   \n",
       "...                                                       ...   \n",
       "AC168977.1                                                0.0   \n",
       "AC149090.1                                                2.0   \n",
       "CAAA01118383.1                                            0.0   \n",
       "Vmn2r122                                                  0.0   \n",
       "CAAA01147332.1                                            0.0   \n",
       "\n",
       "                10X_cells.GGAAAGCGTGCAACGA-L8TX_171026_01_H04  \\\n",
       "Xkr4                                                      5.0   \n",
       "Gm1992                                                    0.0   \n",
       "Gm37381                                                   0.0   \n",
       "Rp1                                                       0.0   \n",
       "Sox17                                                     0.0   \n",
       "...                                                       ...   \n",
       "AC168977.1                                                0.0   \n",
       "AC149090.1                                                8.0   \n",
       "CAAA01118383.1                                            0.0   \n",
       "Vmn2r122                                                  0.0   \n",
       "CAAA01147332.1                                            0.0   \n",
       "\n",
       "                10X_cells.AAGACCTAGTGGCACA-L8TX_190312_01_C02  \\\n",
       "Xkr4                                                      3.0   \n",
       "Gm1992                                                    0.0   \n",
       "Gm37381                                                   0.0   \n",
       "Rp1                                                       0.0   \n",
       "Sox17                                                     0.0   \n",
       "...                                                       ...   \n",
       "AC168977.1                                                0.0   \n",
       "AC149090.1                                                5.0   \n",
       "CAAA01118383.1                                            1.0   \n",
       "Vmn2r122                                                  0.0   \n",
       "CAAA01147332.1                                            0.0   \n",
       "\n",
       "                10X_cells.CCACGGACATCGTCGG-L8TX_180406_01_C06  \\\n",
       "Xkr4                                                      1.0   \n",
       "Gm1992                                                    0.0   \n",
       "Gm37381                                                   0.0   \n",
       "Rp1                                                       0.0   \n",
       "Sox17                                                     0.0   \n",
       "...                                                       ...   \n",
       "AC168977.1                                                0.0   \n",
       "AC149090.1                                                0.0   \n",
       "CAAA01118383.1                                            0.0   \n",
       "Vmn2r122                                                  0.0   \n",
       "CAAA01147332.1                                            0.0   \n",
       "\n",
       "                10X_cells.GAAGCAGCAGCGATCC-L8TX_180221_01_C09  \\\n",
       "Xkr4                                                      4.0   \n",
       "Gm1992                                                    0.0   \n",
       "Gm37381                                                   0.0   \n",
       "Rp1                                                       0.0   \n",
       "Sox17                                                     0.0   \n",
       "...                                                       ...   \n",
       "AC168977.1                                                0.0   \n",
       "AC149090.1                                                0.0   \n",
       "CAAA01118383.1                                            0.0   \n",
       "Vmn2r122                                                  0.0   \n",
       "CAAA01147332.1                                            0.0   \n",
       "\n",
       "                10X_cells.TCAGGTACAGCGTAAG-L8TX_180926_01_C12  \\\n",
       "Xkr4                                                      1.0   \n",
       "Gm1992                                                    0.0   \n",
       "Gm37381                                                   0.0   \n",
       "Rp1                                                       0.0   \n",
       "Sox17                                                     0.0   \n",
       "...                                                       ...   \n",
       "AC168977.1                                                0.0   \n",
       "AC149090.1                                                1.0   \n",
       "CAAA01118383.1                                            0.0   \n",
       "Vmn2r122                                                  0.0   \n",
       "CAAA01147332.1                                            0.0   \n",
       "\n",
       "                10X_cells.CGGACGTCACATGTGT-L8TX_180115_01_G11  \\\n",
       "Xkr4                                                     11.0   \n",
       "Gm1992                                                    0.0   \n",
       "Gm37381                                                   0.0   \n",
       "Rp1                                                       0.0   \n",
       "Sox17                                                     0.0   \n",
       "...                                                       ...   \n",
       "AC168977.1                                                0.0   \n",
       "AC149090.1                                                4.0   \n",
       "CAAA01118383.1                                            1.0   \n",
       "Vmn2r122                                                  0.0   \n",
       "CAAA01147332.1                                            0.0   \n",
       "\n",
       "                10X_cells.TACTTGTCACATTAGC-L8TX_180221_01_E10  \\\n",
       "Xkr4                                                      8.0   \n",
       "Gm1992                                                    0.0   \n",
       "Gm37381                                                   0.0   \n",
       "Rp1                                                       0.0   \n",
       "Sox17                                                     0.0   \n",
       "...                                                       ...   \n",
       "AC168977.1                                                0.0   \n",
       "AC149090.1                                                8.0   \n",
       "CAAA01118383.1                                            0.0   \n",
       "Vmn2r122                                                  0.0   \n",
       "CAAA01147332.1                                            0.0   \n",
       "\n",
       "                10X_cells.CCTTACGGTTCGTGAT-L8TX_180115_01_H10  \n",
       "Xkr4                                                     12.0  \n",
       "Gm1992                                                    0.0  \n",
       "Gm37381                                                   0.0  \n",
       "Rp1                                                       0.0  \n",
       "Sox17                                                     0.0  \n",
       "...                                                       ...  \n",
       "AC168977.1                                                0.0  \n",
       "AC149090.1                                                1.0  \n",
       "CAAA01118383.1                                            0.0  \n",
       "Vmn2r122                                                  0.0  \n",
       "CAAA01147332.1                                            0.0  \n",
       "\n",
       "[31053 rows x 14995 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# shuffle the DataFrame rows\n",
    "RNA_data = RNA_data.sample(frac = 1)\n",
    "\n",
    "RNA_data = RNA_data.T\n",
    "RNA_data = RNA_data.iloc[:,:rna_cells] # downsampling\n",
    "\n",
    "RNA_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# check for low-quality cells\n",
    "# RNA_data[RNA_data.columns[(RNA_data != 0).sum(axis=0) > 2000]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 14995 × 31053"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "RNA_data_adata = sc.AnnData(RNA_data.T)\n",
    "RNA_data_adata"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Spatial data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>2010300C02Rik</th>\n",
       "      <th>Acsbg1</th>\n",
       "      <th>Acta2</th>\n",
       "      <th>Acvrl1</th>\n",
       "      <th>Adamts2</th>\n",
       "      <th>Adamtsl1</th>\n",
       "      <th>Adgrl4</th>\n",
       "      <th>Aldh1a2</th>\n",
       "      <th>Aldh1l1</th>\n",
       "      <th>Angpt1</th>\n",
       "      <th>...</th>\n",
       "      <th>Trpc4</th>\n",
       "      <th>Tubb2a</th>\n",
       "      <th>Unc13c</th>\n",
       "      <th>Vat1l</th>\n",
       "      <th>Vip</th>\n",
       "      <th>Vwc2l</th>\n",
       "      <th>Wfs1</th>\n",
       "      <th>Zfp366</th>\n",
       "      <th>Zfp536</th>\n",
       "      <th>Zfpm2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.672825</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>4.98311</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.364252</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>5.576547</td>\n",
       "      <td>6.267799</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.525453</td>\n",
       "      <td>5.525453</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.216606</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.355722</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>4.355722</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.355722</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.136730</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5.779448</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.089387</td>\n",
       "      <td>5.089387</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.089387</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82936</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.576547</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.267799</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>4.887178</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.887178</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.576547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82937</th>\n",
       "      <td>4.485477</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>4.485477</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82938</th>\n",
       "      <td>3.982138</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.982138</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>3.982138</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.665919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82939</th>\n",
       "      <td>5.492796</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>5.492796</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.402382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82940</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>4.548155</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.548155</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>82941 rows × 284 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       2010300C02Rik  Acsbg1  Acta2    Acvrl1   Adamts2  Adamtsl1    Adgrl4  \\\n",
       "0           0.000000     0.0    0.0  0.000000  0.000000  5.672825  0.000000   \n",
       "1           0.000000     0.0    0.0  0.000000  0.000000  0.000000  0.000000   \n",
       "2           0.000000     0.0    0.0  0.000000  5.525453  5.525453  0.000000   \n",
       "3           0.000000     0.0    0.0  0.000000  0.000000  4.355722  0.000000   \n",
       "4           5.779448     0.0    0.0  0.000000  0.000000  0.000000  0.000000   \n",
       "...              ...     ...    ...       ...       ...       ...       ...   \n",
       "82936       0.000000     0.0    0.0  5.576547  0.000000  0.000000  6.267799   \n",
       "82937       4.485477     0.0    0.0  0.000000  0.000000  0.000000  0.000000   \n",
       "82938       3.982138     0.0    0.0  0.000000  0.000000  0.000000  0.000000   \n",
       "82939       5.492796     0.0    0.0  0.000000  0.000000  0.000000  0.000000   \n",
       "82940       0.000000     0.0    0.0  0.000000  0.000000  0.000000  0.000000   \n",
       "\n",
       "       Aldh1a2  Aldh1l1    Angpt1  ...    Trpc4    Tubb2a    Unc13c     Vat1l  \\\n",
       "0          0.0      0.0  0.000000  ...  4.98311  0.000000  0.000000  0.000000   \n",
       "1          0.0      0.0  0.000000  ...  0.00000  5.576547  6.267799  0.000000   \n",
       "2          0.0      0.0  0.000000  ...  0.00000  0.000000  0.000000  0.000000   \n",
       "3          0.0      0.0  0.000000  ...  0.00000  4.355722  0.000000  4.355722   \n",
       "4          0.0      0.0  0.000000  ...  0.00000  0.000000  5.089387  5.089387   \n",
       "...        ...      ...       ...  ...      ...       ...       ...       ...   \n",
       "82936      0.0      0.0  0.000000  ...  0.00000  4.887178  0.000000  0.000000   \n",
       "82937      0.0      0.0  0.000000  ...  0.00000  4.485477  0.000000  0.000000   \n",
       "82938      0.0      0.0  3.982138  ...  0.00000  3.982138  0.000000  0.000000   \n",
       "82939      0.0      0.0  0.000000  ...  0.00000  5.492796  0.000000  0.000000   \n",
       "82940      0.0      0.0  0.000000  ...  0.00000  4.548155  0.000000  0.000000   \n",
       "\n",
       "       Vip  Vwc2l      Wfs1    Zfp366    Zfp536     Zfpm2  \n",
       "0      0.0    0.0  0.000000  0.000000  6.364252  0.000000  \n",
       "1      0.0    0.0  0.000000  0.000000  0.000000  0.000000  \n",
       "2      0.0    0.0  0.000000  0.000000  6.216606  0.000000  \n",
       "3      0.0    0.0  0.000000  0.000000  6.136730  0.000000  \n",
       "4      0.0    0.0  5.089387  0.000000  0.000000  0.000000  \n",
       "...    ...    ...       ...       ...       ...       ...  \n",
       "82936  0.0    0.0  0.000000  4.887178  0.000000  5.576547  \n",
       "82937  0.0    0.0  0.000000  0.000000  0.000000  0.000000  \n",
       "82938  0.0    0.0  0.000000  0.000000  0.000000  4.665919  \n",
       "82939  0.0    0.0  0.000000  0.000000  0.000000  4.402382  \n",
       "82940  0.0    0.0  0.000000  0.000000  0.000000  4.548155  \n",
       "\n",
       "[82941 rows x 284 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Spatial_data = pd.read_csv(Spatial_file, index_col=0)\n",
    "Spatial_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>2010300C02Rik</th>\n",
       "      <th>Acsbg1</th>\n",
       "      <th>Acta2</th>\n",
       "      <th>Acvrl1</th>\n",
       "      <th>Adamts2</th>\n",
       "      <th>Adamtsl1</th>\n",
       "      <th>Adgrl4</th>\n",
       "      <th>Aldh1a2</th>\n",
       "      <th>Aldh1l1</th>\n",
       "      <th>Angpt1</th>\n",
       "      <th>...</th>\n",
       "      <th>Trpc4</th>\n",
       "      <th>Tubb2a</th>\n",
       "      <th>Unc13c</th>\n",
       "      <th>Vat1l</th>\n",
       "      <th>Vip</th>\n",
       "      <th>Vwc2l</th>\n",
       "      <th>Wfs1</th>\n",
       "      <th>Zfp366</th>\n",
       "      <th>Zfp536</th>\n",
       "      <th>Zfpm2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.726633</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.747799</td>\n",
       "      <td>5.747799</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.747799</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.844283</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.880932</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.787889</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.940848</td>\n",
       "      <td>4.940848</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>4.940848</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.321767</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>5.948467</td>\n",
       "      <td>4.855061</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.169672</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8289</th>\n",
       "      <td>5.408168</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.812139</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.719491</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8290</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.968821</td>\n",
       "      <td>5.658486</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.282601</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.658486</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8291</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.603116</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.294418</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8292</th>\n",
       "      <td>4.874206</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.874206</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8293</th>\n",
       "      <td>4.764556</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.764556</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>4.764556</td>\n",
       "      <td>5.857468</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.764556</td>\n",
       "      <td>4.764556</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8294 rows × 284 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      2010300C02Rik  Acsbg1  Acta2  Acvrl1   Adamts2  Adamtsl1    Adgrl4  \\\n",
       "0          0.000000     0.0    0.0     0.0  0.000000  0.000000  6.726633   \n",
       "1          0.000000     0.0    0.0     0.0  5.747799  5.747799  0.000000   \n",
       "2          0.000000     0.0    0.0     0.0  0.000000  0.000000  0.000000   \n",
       "3          0.000000     0.0    0.0     0.0  4.940848  4.940848  0.000000   \n",
       "4          0.000000     0.0    0.0     0.0  0.000000  0.000000  0.000000   \n",
       "...             ...     ...    ...     ...       ...       ...       ...   \n",
       "8289       5.408168     0.0    0.0     0.0  0.000000  0.000000  0.000000   \n",
       "8290       0.000000     0.0    0.0     0.0  0.000000  4.968821  5.658486   \n",
       "8291       0.000000     0.0    0.0     0.0  0.000000  5.603116  0.000000   \n",
       "8292       4.874206     0.0    0.0     0.0  0.000000  0.000000  0.000000   \n",
       "8293       4.764556     0.0    0.0     0.0  0.000000  4.764556  0.000000   \n",
       "\n",
       "      Aldh1a2  Aldh1l1  Angpt1  ...     Trpc4    Tubb2a  Unc13c  Vat1l  Vip  \\\n",
       "0         0.0      0.0     0.0  ...  0.000000  0.000000     0.0    0.0  0.0   \n",
       "1         0.0      0.0     0.0  ...  0.000000  5.747799     0.0    0.0  0.0   \n",
       "2         0.0      0.0     0.0  ...  0.000000  5.880932     0.0    0.0  0.0   \n",
       "3         0.0      0.0     0.0  ...  4.940848  0.000000     0.0    0.0  0.0   \n",
       "4         0.0      0.0     0.0  ...  5.948467  4.855061     0.0    0.0  0.0   \n",
       "...       ...      ...     ...  ...       ...       ...     ...    ...  ...   \n",
       "8289      0.0      0.0     0.0  ...  0.000000  0.000000     0.0    0.0  0.0   \n",
       "8290      0.0      0.0     0.0  ...  0.000000  4.282601     0.0    0.0  0.0   \n",
       "8291      0.0      0.0     0.0  ...  0.000000  0.000000     0.0    0.0  0.0   \n",
       "8292      0.0      0.0     0.0  ...  0.000000  0.000000     0.0    0.0  0.0   \n",
       "8293      0.0      0.0     0.0  ...  4.764556  5.857468     0.0    0.0  0.0   \n",
       "\n",
       "         Vwc2l      Wfs1  Zfp366    Zfp536     Zfpm2  \n",
       "0     0.000000  0.000000     0.0  0.000000  0.000000  \n",
       "1     0.000000  0.000000     0.0  6.844283  0.000000  \n",
       "2     4.787889  0.000000     0.0  0.000000  0.000000  \n",
       "3     0.000000  0.000000     0.0  6.321767  0.000000  \n",
       "4     0.000000  0.000000     0.0  4.169672  0.000000  \n",
       "...        ...       ...     ...       ...       ...  \n",
       "8289  0.000000  5.812139     0.0  0.000000  4.719491  \n",
       "8290  0.000000  0.000000     0.0  5.658486  0.000000  \n",
       "8291  0.000000  0.000000     0.0  6.294418  0.000000  \n",
       "8292  4.874206  0.000000     0.0  0.000000  0.000000  \n",
       "8293  0.000000  0.000000     0.0  4.764556  4.764556  \n",
       "\n",
       "[8294 rows x 284 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# shuffle the DataFrame rows\n",
    "Spatial_data = Spatial_data.sample(frac = 1).reset_index(drop=True)\n",
    "\n",
    "Spatial_data = Spatial_data.iloc[:spatial_spots,:] # downsampling\n",
    "Spatial_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save to calculate metrics\n",
    "\n",
    "# Spatial_data.to_csv(outdir + '/spatial_data.csv', header = 1, index = 1)\n",
    "# print(outdir + '/spatial_data.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 8294 × 284"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Spatial_data_adata = sc.AnnData(Spatial_data)\n",
    "Spatial_data_adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(8294, 2)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "locations = np.array(pd.read_csv(location_file, index_col=0))[:spatial_spots] #while testing\n",
    "locations.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Training and testing list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total genes:  282\n",
      "test genes:  28\n",
      "0\n",
      "1\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "train list shape:  (10, 254)\n",
      "test list shape:  (10, 28)\n"
     ]
    }
   ],
   "source": [
    "import random \n",
    "\n",
    "RNA_data_train = RNA_data\n",
    "\n",
    "RNA_data_train = RNA_data_train.loc[(RNA_data_train.sum(axis=1) != 0)] # avoiding error in SpaGE\n",
    "RNA_data_train = RNA_data_train.loc[(RNA_data_train.var(axis=1) != 0)]\n",
    "\n",
    "Spatial_data_cross = list(RNA_data_train.index&Spatial_data.columns) #list(Spatial_data.columns)\n",
    "\n",
    "n_total = len(Spatial_data_cross)\n",
    "print('total genes: ', n_total)\n",
    "\n",
    "n_test = int(len(Spatial_data_cross)/fraction) # floor\n",
    "print('test genes: ', n_test)\n",
    "\n",
    "train_gene = []\n",
    "predict_gene = []\n",
    "\n",
    "for i in range(iterations):\n",
    "    print(i)\n",
    "\n",
    "    random.shuffle(Spatial_data_cross)\n",
    "\n",
    "    Spatial_data_test = Spatial_data_cross[:n_test]\n",
    "    Spatial_data_train = Spatial_data_cross[n_test:]\n",
    "\n",
    "    train_gene.append(Spatial_data_train)\n",
    "    predict_gene.append(Spatial_data_test)\n",
    "    \n",
    "\n",
    "print('train list shape: ', np.array(train_gene).shape)\n",
    "print('test list shape: ', np.array(predict_gene).shape)\n",
    "\n",
    " # save lists\n",
    "train_gene_df = pd.DataFrame(train_gene)\n",
    "test_gene_df = pd.DataFrame(predict_gene)\n",
    "\n",
    "# train_gene_df.to_csv(outdir + '/train_list.csv')\n",
    "# test_gene_df.to_csv(outdir + '/test_list.csv')\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Predicting undetected transcripts by each method"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SpaGE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "xenium data/Figure2/downsampling_10/iterations_10/split_10/impute/SpaGE\n",
      "[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]\n",
      "We run SpaGE for this data\n",
      "We run SpaGE for this data\n",
      "We run SpaGE for this data\n",
      "We run SpaGE for this data\n",
      "We run SpaGE for this data\n",
      "We run SpaGE for this data\n",
      "\n",
      "\n",
      "We run SpaGE for this data\n",
      "We run SpaGE for this data\n",
      "We run SpaGE for this data\n",
      "We run SpaGE for this data\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "train gene\n",
      "train gene\n",
      "train gene\n",
      "train gene\n",
      "train gene\n",
      "train gene\n",
      "train gene\n",
      "train genetrain gene\n",
      "\n",
      "train gene\n",
      "Img_Genes\n",
      "Img_Genes\n",
      "Img_Genes\n",
      "Img_Genes\n",
      "Img_Genes\n",
      "Img_Genes\n",
      "Img_Genes\n",
      "Img_Genes\n",
      "Img_Genes\n",
      "Img_Genes\n",
      "saved file:  xenium data/Figure2/downsampling_10/iterations_10/split_10/impute/SpaGE/SpaGE_impute.csv\n"
     ]
    }
   ],
   "source": [
    "if 'spage' in methods:\n",
    "    \n",
    "    savedir = outdir + '/SpaGE'\n",
    "\n",
    "    if not os.path.exists(savedir):\n",
    "        os.mkdir(savedir)\n",
    "\n",
    "    print(savedir)\n",
    "\n",
    "    train_gene_df.to_csv(savedir + '/train_list.csv')\n",
    "    test_gene_df.to_csv(savedir + '/test_list.csv')\n",
    "\n",
    "    Spatial_data.to_csv(savedir + '/spatial_data.csv', header = 1, index = 1)\n",
    "    \n",
    "    KFOLD = list(range(len(train_gene)))\n",
    "    \n",
    "    print(KFOLD)\n",
    "    \n",
    "    with multiprocessing.Pool(10) as pool:\n",
    "        result = pd.concat(pool.map(SpaGE_impute, KFOLD), axis = 1)\n",
    "    result.to_csv(savedir + '/SpaGE_impute'  + '.csv', header = 1, index = 1)\n",
    "    print('saved file: ', savedir + '/SpaGE_impute'  + '.csv')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SpaOTsc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### novoSpaRc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GPU Platform gimVI, Tangram, and stPlus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### gimVI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if 'gimvi' in methods:\n",
    "\n",
    "    savedir = outdir + '/gimVI'\n",
    "\n",
    "    if not os.path.exists(savedir):\n",
    "        os.mkdir(savedir)\n",
    "\n",
    "    train_gene_df.to_csv(savedir + '/train_list.csv')\n",
    "    test_gene_df.to_csv(savedir + '/test_list.csv')\n",
    "\n",
    "    Spatial_data.to_csv(savedir + '/spatial_data.csv', header = 1, index = 1)\n",
    "    \n",
    "    KFOLD = list(range(len(train_gene)))\n",
    "    with multiprocessing.Pool(10) as pool:\n",
    "        result = pd.concat(pool.map(gimVI_impute, KFOLD), axis = 1)\n",
    "    result.to_csv(savedir + '/gimVI_impute'  + '.csv', header = 1, index = 1)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tangram"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We run Tangram for this \n",
      "xenium data/We run Tangram for this \n",
      "xenium data/We run Tangram for this \n",
      "xenium data/We run Tangram for this \n",
      "xenium data/We run Tangram for this \n",
      "xenium data/We run Tangram for this \n",
      "xenium data/We run Tangram for this \n",
      "xenium data/We run Tangram for this \n",
      "xenium data/We run Tangram for this \n",
      "xenium data/We run Tangram for this \n",
      "xenium data/\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:254 training genes are saved in `uns``training_genes` of both single cell and spatial Anndatas.\n",
      "INFO:root:282 overlapped genes are saved in `uns``overlap_genes` of both single cell and spatial Anndatas.\n",
      "INFO:root:uniform based density prior is calculated and saved in `obs``uniform_density` of the spatial Anndata.\n",
      "INFO:root:rna count based density prior is calculated and saved in `obs``rna_count_based_density` of the spatial Anndata.\n",
      "INFO:root:Allocate tensors for mapping.\n",
      "INFO:root:Begin training with 254 genes and rna_count_based density_prior in clusters mode...\n",
      "INFO:root:254 training genes are saved in `uns``training_genes` of both single cell and spatial Anndatas.\n",
      "INFO:root:282 overlapped genes are saved in `uns``overlap_genes` of both single cell and spatial Anndatas.\n",
      "INFO:root:uniform based density prior is calculated and saved in `obs``uniform_density` of the spatial Anndata.\n",
      "INFO:root:rna count based density prior is calculated and saved in `obs``rna_count_based_density` of the spatial Anndata.\n",
      "INFO:root:Allocate tensors for mapping.\n",
      "INFO:root:Begin training with 254 genes and rna_count_based density_prior in clusters mode...\n",
      "INFO:root:254 training genes are saved in `uns``training_genes` of both single cell and spatial Anndatas.\n",
      "INFO:root:282 overlapped genes are saved in `uns``overlap_genes` of both single cell and spatial Anndatas.\n",
      "INFO:root:uniform based density prior is calculated and saved in `obs``uniform_density` of the spatial Anndata.\n",
      "INFO:root:rna count based density prior is calculated and saved in `obs``rna_count_based_density` of the spatial Anndata.\n",
      "INFO:root:Allocate tensors for mapping.\n",
      "INFO:root:Begin training with 254 genes and rna_count_based density_prior in clusters mode...\n",
      "INFO:root:254 training genes are saved in `uns``training_genes` of both single cell and spatial Anndatas.\n",
      "INFO:root:282 overlapped genes are saved in `uns``overlap_genes` of both single cell and spatial Anndatas.\n",
      "INFO:root:uniform based density prior is calculated and saved in `obs``uniform_density` of the spatial Anndata.\n",
      "INFO:root:rna count based density prior is calculated and saved in `obs``rna_count_based_density` of the spatial Anndata.\n",
      "INFO:root:254 training genes are saved in `uns``training_genes` of both single cell and spatial Anndatas.\n",
      "INFO:root:282 overlapped genes are saved in `uns``overlap_genes` of both single cell and spatial Anndatas.\n",
      "INFO:root:uniform based density prior is calculated and saved in `obs``uniform_density` of the spatial Anndata.\n",
      "INFO:root:rna count based density prior is calculated and saved in `obs``rna_count_based_density` of the spatial Anndata.\n",
      "INFO:root:Printing scores every 100 epochs.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score: 0.311, KL reg: 0.135\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Allocate tensors for mapping.\n",
      "INFO:root:Begin training with 254 genes and rna_count_based density_prior in clusters mode...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score: 0.569, KL reg: 0.001\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Allocate tensors for mapping.\n",
      "INFO:root:Begin training with 254 genes and rna_count_based density_prior in clusters mode...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score: 0.571, KL reg: 0.001\n",
      "Score: 0.571, KL reg: 0.001\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:254 training genes are saved in `uns``training_genes` of both single cell and spatial Anndatas.\n",
      "INFO:root:282 overlapped genes are saved in `uns``overlap_genes` of both single cell and spatial Anndatas.\n",
      "INFO:root:uniform based density prior is calculated and saved in `obs``uniform_density` of the spatial Anndata.\n",
      "INFO:root:rna count based density prior is calculated and saved in `obs``rna_count_based_density` of the spatial Anndata.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score: 0.572, KL reg: 0.001\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:254 training genes are saved in `uns``training_genes` of both single cell and spatial Anndatas.\n",
      "INFO:root:282 overlapped genes are saved in `uns``overlap_genes` of both single cell and spatial Anndatas.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score: 0.572, KL reg: 0.001\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:uniform based density prior is calculated and saved in `obs``uniform_density` of the spatial Anndata.\n",
      "INFO:root:rna count based density prior is calculated and saved in `obs``rna_count_based_density` of the spatial Anndata.\n",
      "INFO:root:Allocate tensors for mapping.\n",
      "INFO:root:Begin training with 254 genes and rna_count_based density_prior in clusters mode...\n",
      "INFO:root:Printing scores every 100 epochs.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score: 0.306, KL reg: 0.135\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:254 training genes are saved in `uns``training_genes` of both single cell and spatial Anndatas.\n",
      "INFO:root:282 overlapped genes are saved in `uns``overlap_genes` of both single cell and spatial Anndatas.\n",
      "INFO:root:uniform based density prior is calculated and saved in `obs``uniform_density` of the spatial Anndata.\n",
      "INFO:root:rna count based density prior is calculated and saved in `obs``rna_count_based_density` of the spatial Anndata.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score: 0.572, KL reg: 0.001\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Allocate tensors for mapping.\n",
      "INFO:root:Begin training with 254 genes and rna_count_based density_prior in clusters mode...\n",
      "INFO:root:254 training genes are saved in `uns``training_genes` of both single cell and spatial Anndatas.\n",
      "INFO:root:254 training genes are saved in `uns``training_genes` of both single cell and spatial Anndatas.\n",
      "INFO:root:282 overlapped genes are saved in `uns``overlap_genes` of both single cell and spatial Anndatas.\n",
      "INFO:root:282 overlapped genes are saved in `uns``overlap_genes` of both single cell and spatial Anndatas.\n",
      "INFO:root:uniform based density prior is calculated and saved in `obs``uniform_density` of the spatial Anndata.\n",
      "INFO:root:uniform based density prior is calculated and saved in `obs``uniform_density` of the spatial Anndata.\n",
      "INFO:root:rna count based density prior is calculated and saved in `obs``rna_count_based_density` of the spatial Anndata.\n",
      "INFO:root:rna count based density prior is calculated and saved in `obs``rna_count_based_density` of the spatial Anndata.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score: 0.564, KL reg: 0.001\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Allocate tensors for mapping.\n",
      "INFO:root:Begin training with 254 genes and rna_count_based density_prior in clusters mode...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score: 0.572, KL reg: 0.001\n",
      "Score: 0.567, KL reg: 0.001\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Allocate tensors for mapping.\n",
      "INFO:root:Allocate tensors for mapping.\n",
      "INFO:root:Begin training with 254 genes and rna_count_based density_prior in clusters mode...INFO:root:Begin training with 254 genes and rna_count_based density_prior in clusters mode...\n",
      "\n",
      "INFO:root:Printing scores every 100 epochs.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score: 0.302, KL reg: 0.135\n",
      "Score: 0.572, KL reg: 0.001\n",
      "Score: 0.567, KL reg: 0.001\n",
      "Score: 0.559, KL reg: 0.001\n",
      "Score: 0.572, KL reg: 0.001\n",
      "Score: 0.567, KL reg: 0.001\n",
      "Score: 0.561, KL reg: 0.001\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Saving results..\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score: 0.568, KL reg: 0.001\n",
      "Score: 0.562, KL reg: 0.001\n",
      "Score: 0.568, KL reg: 0.001\n",
      "Score: 0.562, KL reg: 0.001\n",
      "Score: 0.568, KL reg: 0.001\n",
      "Score: 0.562, KL reg: 0.001\n",
      "Score: 0.568, KL reg: 0.001\n",
      "Score: 0.562, KL reg: 0.001\n",
      "Score: 0.568, KL reg: 0.001\n",
      "Score: 0.563, KL reg: 0.001\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Saving results..\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score: 0.563, KL reg: 0.001\n",
      "Score: 0.563, KL reg: 0.001\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Saving results..\n"
     ]
    }
   ],
   "source": [
    "if 'tangram' in methods:\n",
    "    \n",
    "    savedir = outdir + '/Tangram'\n",
    "\n",
    "    if not os.path.exists(savedir):\n",
    "        os.mkdir(savedir)\n",
    "\n",
    "    train_gene_df.to_csv(savedir + '/train_list.csv')\n",
    "    test_gene_df.to_csv(savedir + '/test_list.csv')\n",
    "\n",
    "    Spatial_data.to_csv(savedir + '/spatial_data.csv', header = 1, index = 1)\n",
    "    \n",
    "    KFOLD = list(range(len(train_gene)))\n",
    "    with multiprocessing.Pool(10) as pool:\n",
    "        result = pd.concat(pool.map(Tangram_impute, KFOLD), axis = 1)\n",
    "    result.to_csv(savedir + '/Tangram_impute'  + '.csv', header = 1, index = 1)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CPU slow"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### novosparc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "We run novoSpaRc for this data\n",
      "We run novoSpaRc for this data\n",
      "We run novoSpaRc for this data\n",
      "We run novoSpaRc for this data\n",
      "\n",
      "We run novoSpaRc for this data\n",
      "We run novoSpaRc for this data\n",
      "We run novoSpaRc for this data\n",
      "\n",
      "We run novoSpaRc for this data\n",
      "We run novoSpaRc for this data\n",
      "We run novoSpaRc for this data\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "number of cells and genes in the matrix:number of cells and genes in the matrix:number of cells and genes in the matrix:number of cells and genes in the matrix:"
     ]
    }
   ],
   "source": [
    "if 'novosparc' in methods:\n",
    "    savedir = outdir + '/novoSpaRc'\n",
    "\n",
    "    if not os.path.exists(savedir):\n",
    "        os.mkdir(savedir)\n",
    "\n",
    "    train_gene_df.to_csv(savedir + '/train_list.csv')\n",
    "    test_gene_df.to_csv(savedir + '/test_list.csv')\n",
    "\n",
    "    Spatial_data.to_csv(savedir + '/spatial_data.csv', header = 1, index = 1)\n",
    "    \n",
    "    KFOLD = list(range(len(train_gene)))\n",
    "    with multiprocessing.Pool(10) as pool:\n",
    "        result = pd.concat(pool.map(novoSpaRc_impute, KFOLD), axis = 1)\n",
    "    result.to_csv(savedir + '/novoSpaRc_impute'  + '.csv', header = 1, index = 1)\n",
    "    print('saved file: ', savedir + '/novorSpaRC_impute'  + '.csv')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Spaotsc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if 'spaotsc' in methods:\n",
    "    \n",
    "    savedir = outdir + '/SpaOTsc'\n",
    "\n",
    "    if not os.path.exists(savedir):\n",
    "        os.mkdir(savedir)\n",
    "\n",
    "    train_gene_df.to_csv(savedir + '/train_list.csv')\n",
    "    test_gene_df.to_csv(savedir + '/test_list.csv')\n",
    "\n",
    "    Spatial_data.to_csv(savedir + '/spatial_data.csv', header = 1, index = 1)\n",
    "    \n",
    "    KFOLD = list(range(len(train_gene)))\n",
    "    print(KFOLD)\n",
    "    \n",
    "    with multiprocessing.Pool(10) as pool:\n",
    "        result = pd.concat(pool.map(SpaOTsc_impute, KFOLD), axis = 1)\n",
    "        \n",
    "    result.to_csv(savedir + '/SpaOTsc_impute'  + '.csv', header = 1, index = 1)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### stPlus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from stPlus import *\n",
    "# KFOLD = list(range(len(train_gene)))\n",
    "result = pd.DataFrame()\n",
    "for n in KFOLD:\n",
    "    tmp = stPlus_impute(n)\n",
    "    result = pd.concat([result, tmp], axis=1)\n",
    "result.to_csv(outdir +  '/stPlus_impute'  + '.csv',header = 1, index = 1)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Seurat impute file process"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This outdir is none for Seurat : FigureData/Figure2/Dataset4\n"
     ]
    }
   ],
   "source": [
    "# input_path = 'DataUpload/Dataset4/'\n",
    "# output_path = 'FigureData/Figure2/Dataset4/'\n",
    "# os.system('Rscript SpatialBenchmarking/FigureData/Figure2/RCodes/Seurat.r ' + input_path + ' ' + output_path)\n",
    "\n",
    "# PATH = 'FigureData/Figure2/'\n",
    "# DataSets = ['Dataset4']\n",
    "# for Data in DataSets:\n",
    "#     impute_count_file = PATH + Data + '/Seurat_impute.csv'\n",
    "#     if os.path.exists(impute_count_file):\n",
    "#         df = pd.read_csv(impute_count_file, header = 0, index_col = 0)\n",
    "#         if df.index[0] == 0:\n",
    "#             A = df\n",
    "#             continue\n",
    "#         print (impute_count_file)\n",
    "#         Index = [(int(x)-1) for x in df.index]\n",
    "#         Index = sorted(Index)\n",
    "#         df.index = Index\n",
    "#         B = df\n",
    "#         df.to_csv(impute_count_file)\n",
    "#     else:\n",
    "#         print ('This outdir is none for Seurat : ' + PATH + Data)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# LIGER impute file process"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# input_path = 'DataUpload/Dataset4/'\n",
    "# output_path = 'SpatialBenchmarking/FigureData/Figure2/Dataset4/'\n",
    "\n",
    "# os.system('Rscript SpatialBenchmarking/FigureData/Figure2/RCodes/LIGER.r'  + 'DataUpload/Dataset4/' + ' ' + 'SpatialBenchmarking/FigureData/Figure2/Dataset4/')\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This outdir is none for LIGER : SpatialBenchmarking/FigureData/Figure2/Dataset4/Dataset4\n"
     ]
    }
   ],
   "source": [
    "# input_path = 'DataUpload/Dataset4/'\n",
    "# output_path = 'SpatialBenchmarking/FigureData/Figure2/Dataset4/'\n",
    "# os.system('Rscript SpatialBenchmarking/FigureData/Figure2/RCodes/LIGER.r ' + input_path + ' ' + output_path)\n",
    "\n",
    "# PATH = 'SpatialBenchmarking/FigureData/Figure2/Dataset4/'\n",
    "# DataSets = ['Dataset4']\n",
    "# for Data in DataSets:\n",
    "#     impute_count_file = PATH + Data + '/LIGER_impute.csv'\n",
    "#     if os.path.exists(impute_count_file):\n",
    "#         df = pd.read_csv(impute_count_file, header = 0, index_col = 0)\n",
    "#         if df.index[0] == 0:\n",
    "#             A = df\n",
    "#             continue\n",
    "#         print (impute_count_file)\n",
    "#         Index = [(int(x.replace('V', '')) -1) for x in df.index]\n",
    "#         Index = sorted(Index)\n",
    "#         df.index = Index\n",
    "#         B = df\n",
    "#         df.to_csv(impute_count_file)\n",
    "#     else:\n",
    "#         print ('This outdir is none for LIGER : ' + PATH + Data)\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Benchmarking",
   "language": "python",
   "name": "benchmarking"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
